RuleTool.py

```diff
@@ -131,6 +108,41 @@ def reconstruct_rule(rule: dict) -> str:
     return rule_text
 
 
+# Procedural Functions
+def read_file(file: str) -> str:
+    if not (file and pathlib.Path(file).is_file()):
+        print("Invalid Path")
+        exit(1)
+    file_text = pathlib.Path(file).read_text()
+    return file_text
+
+
+def extract_rule_metadata(file_text: str, requested_actions: list) -> dict:
+    text = ""
+    rules = []
+    while file_text.find("(0:") > -1:
+        rule = enumerate_rules(file_text)
+        file_text = file_text[rule["rule_end"]:]
+        if not rule["text"]:
+            continue
+        rule = separate_rule_components(rule)
+        if not rule["action"]:
+            continue
+        rule = separate_action_components(rule)
+        rule = separate_action_subcomponents(rule)
+        if not determine_requested_actions(rule, requested_actions):
+            continue
+        rule_text = reconstruct_rule(rule)
```

Comment:
Can you store a separate original dictionary and skip this step of restoring each rule?